File: SonarApp/app/src/main/java/com/golovkobalak/sonarapp/service/TrackingService.java
package com.golovkobalak.sonarapp.service;

import android.util.Log;

import com.golovkobalak.sonarapp.SonarContext;
import com.golovkobalak.sonarapp.model.GeoSquare;
import com.golovkobalak.sonarapp.model.Marker;
import com.golovkobalak.sonarapp.model.SonarData;
import com.golovkobalak.sonarapp.repository.SonarDataRepository;
import com.google.gson.Gson;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class TrackingService {
    private static final String TAG = TrackingService.class.getSimpleName();
    private static final Gson gson = new Gson();
    private static final SonarDataRepository repo = new SonarDataRepository();

    public void saveTrackingList(String data) {
        Log.i(TAG, "saveTrackingList: " + data);
        try {
            final SonarData[] sonarDataArray = gson.fromJson(data, SonarData[].class);
            repo.saveList(Arrays.asList(sonarDataArray));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public String getMapCacheDir() {
        return "file://" + SonarContext.FILES_DIR_ABS_PATH + "/Tiles";
    }

    public List<Marker> getMarkers(GeoSquare geoSquare) {
        final ArrayList<Marker> list = new ArrayList<>();
        try {
            final List<SonarData> markers = repo.findByGeoSquare(geoSquare);
            for (SonarData marker : markers) {
                list.add(new Marker(marker.getDepth(), marker.getLatitude(), marker.getLongitude()));
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return list;
    }
}

File: SonarApp/app/src/main/java/com/golovkobalak/sonarapp/service/TrackingService.kt
package com.golovkobalak.sonarapp.service

import android.util.Log
import com.golovkobalak.sonarapp.SonarContext.filesDirAbsPath
import com.golovkobalak.sonarapp.model.GeoSquare
import com.golovkobalak.sonarapp.model.Marker
import com.golovkobalak.sonarapp.model.SonarData
import com.golovkobalak.sonarapp.repository.RepoFactory
import com.google.gson.Gson
import java.util.*

class TrackingService {
    fun saveTrackingList(data: String) {
        Log.i(TAG, "saveTrackingList: $data")
        try {
            val sonarDataArray = gson.fromJson(data, Array<SonarData>::class.java)
            repo.saveList(Arrays.asList(*sonarDataArray))
        } catch (e: Exception) {
            Log.w(this.javaClass.name, e)
        }
    }

    val mapCacheDir: String
        get() = "file://" + filesDirAbsPath + "/Tiles"

    fun getMarkers(geoSquare: GeoSquare?): List<Marker> {
        val list = ArrayList<Marker>()
        try {
            val markers = repo.findByGeoSquare(geoSquare)
            for (marker in markers) {
                list.add(Marker(marker.depth!!, marker.latitude, marker.longitude))
            }
        } catch (e: Exception) {
            Log.w(this.javaClass.name, e)
        }
        return list
    }

    fun save(message: String) {
        try {
            Log.i(TAG, "save: $message")
            try {
                val sonarData = gson.fromJson(message, SonarData::class.java)
                if (LocationHelper.getLastLocation() != null) {
                    sonarData.latitude = LocationHelper.getLastLocation().latitude
                    sonarData.longitude = LocationHelper.getLastLocation().longitude
                    sonarData.accuracy = LocationHelper.getLastLocation().accuracy.toString()
                    sonarData.speed = LocationHelper.getLastLocation().speed.toString()
                }

                repo.save(sonarData)
            } catch (e: Exception) {
                Log.w(this.javaClass.name, e)
            }
        } catch (e: Exception) {
            Log.w(this.javaClass.name, e)
        }
    }

    companion object {
        private val TAG = TrackingService::class.java.simpleName
        private val gson = Gson()
        private val repo = RepoFactory().sonarDataRepository();
    }
}